Documentation for AIA (#26940)
keycloak/keycloak: docs/documentation/server_admin/topics/users/con-aia.adoc

// Module included in the following assemblies:
//
// server_admin/topics/users.adoc

[id="con-aia_{context}"]
= Application initiated actions

> Application initiated actions (AIA) allow client application to request user to perform some action on {project_name} side. Usually when OIDC client application

Reviewer: ```suggestion
Application initiated actions (AIA) allow client applications to request a user to perform an action on the {project_name} side. Usually, when an OIDC client application
```

> wants user to login, it redirects him to the login URL as described in the <<con-oidc_{context}, OIDC section>>. After login, user is redirected back to the client application.

Reviewer: ```suggestion
wants a user to login, it redirects that user to the login URL as described in the <<con-oidc_{context}, OIDC section>>. After login, the user is redirected back to the client application.
```

Reviewer: Avoid he, him, his.

He needs to optionally perform the actions, which were added to his account by administrator as described in the <<proc-setting-required-actions_{context}, previous section>>

> and then being redirected back to the application immediately. However AIA allows client application to request some required actions from the user during login. This can be

Reviewer: ```suggestion
and then is immediately redirected back to the application. However, AIA allows the client application to request some required actions from the user during login. This can be
```

> done even if user is already authenticated on the client and has existing SSO session. It is triggered by adding `kc_action` parameter to the OIDC login URL with value containing the requested action.

Reviewer: ```suggestion
done even if the user is already authenticated on the client and has an active SSO session. It is triggered by adding the `kc_action` parameter to the OIDC login URL with the value containing the requested action.
```

For instance `kc_action=UPDATE_PASSWORD` parameter.

> NOTE: The `kc_action` parameter is {project_name} proprietary mechanism unsupported by the OIDC specification.

Reviewer: ```suggestion
NOTE: The `kc_action` parameter is a {project_name} proprietary mechanism unsupported by the OIDC specification.
```

NOTE: Application initiated actions are supported only for OIDC clients.

> So if AIA is used, the example flow is like this:

Reviewer: ```suggestion
So if AIA is used, an example flow is similar to the following:
```

* Client application redirects user to the OIDC login URL with the additional added parameter like for instance `kc_action=UPDATE_PASSWORD`

> * There is `browser` flow always triggered as described in the <<_authentication-flows, Authentication flows section>>. In case that user was not authenticated, he needs to authenticate as during normal login.

Reviewer: ```suggestion
* There is a `browser` flow always triggered as described in the <<_authentication-flows, Authentication flows section>>. If the user was not authenticated, that user needs to authenticate as during normal login.
```

> In case user was already authenticated before, he might be automatically re-authenticated by SSO cookie without a need to actively re-authenticate and fill his credentials again. In this case, he will be

Reviewer: ```suggestion
In case the user was already authenticated, that user might be automatically re-authenticated by an SSO cookie without needing to actively re-authenticate and supply the credentials again. In this case, that user will be
```

directly redirected to the screen with the particular action (update password in this case). But note that in some cases, active re-authentication is required even if user has SSO
cookie (See <<con-aia-reauth_{context}, below>> for the details).

> * The screen with particular action (in this case `update password`) is displayed to the user, so he needs to perform particular action

Reviewer: ```suggestion
* The screen with particular action (in this case `update password`) is displayed to the user, so that user needs to perform a particular action
```

* Then user is redirected back to the client application

> Note that AIA are used by {project_name} own <<_account-service, account console client application>> to request update password or reset other credentials like for instance OTP or WebAuthn.

Reviewer: ```suggestion
Note that AIA are used by the {project_name} <<_account-service, Account Console>> to request update password or to reset other credentials such as OTP or WebAuthn.
```

> WARNING: Even if parameter `kc_action` was used, it is not sufficient to assume that user always performs the action. There is always a chance that user manually deleted

Reviewer: ```suggestion
WARNING: Even if the parameter `kc_action` was used, it is not sufficient to assume that the user always performs the action.  For example, a user could have manually deleted
```

> the `kc_action` parameter from the browser URL. For instance it is not safe to assume that user must have OTP on his account after client requested `kc_action=CONFIGURE_TOTP`. If you

Reviewer: ```suggestion
the `kc_action` parameter from the browser URL. Therefore, no guarantee exists that the user OTP on the account after the client requested `kc_action=CONFIGURE_TOTP`. If you
```

> want to make sure that user really configured 2nd factor authenticator, it may be also needed that client application checks by some other way that user used 2nd factor authentication. For instance

Reviewer: ```suggestion
want to verify that the user configured two-factor authenticator, the client application may need to check it was configured. For instance
```

by checking the claims like `acr` in the tokens.

[id="con-aia-reauth_{context}"]
== Re-authentication during AIA

In case user is already authenticated due the existing SSO session, he usually does not need to actively re-authenticate again. However in case that user actively authenticated longer than 5 minutes ago,

> he will be still requested to re-authenticate when some AIA is requested. There are exceptions from this as described below:

Reviewer: ```suggestion
the client can  still request re-authenticate when some AIA is requested. Exceptions exist from this guideline as follows:

> * The action `delete_account` will always require user to actively re-authenticate

Reviewer: ```suggestion
* The action `delete_account` will always require the user to actively re-authenticate
```

> * The action `update_password` might require user to actively re-authenticate according to the configured <<maximum-authentication-age, Maximum Authentication Age Password policy>>.

Reviewer: ```suggestion
* The action `update_password` might require the user to actively re-authenticate according to the configured <<maximum-authentication-age, Maximum Authentication Age Password policy>>.
```

> In case policy is not configured, it also defaults to 5 minutes.

Reviewer: ```suggestion
In case the policy is not configured, it also defaults to 5 minutes.
```

Reviewer: We always spell out numbers 0 to 9.

> * If you want to use shorter re-authentication, you can still use parameter query parameter like `max_age` with the specified shorter value or eventually `prompt=login`, which will always require user to

Reviewer: ```suggestion
* If you want to use a shorter re-authentication, you can still use a parameter query parameter such as  `max_age` with the specified shorter value or eventually `prompt=login`, which will always require user to
```

> actively re-authenticate as described in the OIDC specification. Note that using `max_age` for longer value than the default 5 minutes (or the one prescribed by password policy) is not supported.

Reviewer: ```suggestion
actively re-authenticate as described in the OIDC specification. Note that using `max_age` for a longer value than the default five minutes (or the one prescribed by password policy) is not supported.
```

> The `max_age` can be currently used only to make the value shorter than the default 5 minutes.

Reviewer: ```suggestion
The `max_age` can be currently used only to make the value shorter than the default five minutes.
```

[id="con-aia-available-actions_{context}"]
== Available actions

> To see all available actions, it can be good to login to the admin console and then in the right top corner click `Realm info` -> tab `Provider info` -> Find provider `required-action` .

Reviewer: ```suggestion
To see all available actions, log in to the Admin Console and got to the top right top corner to click `Realm info` -> tab `Provider info` -> Find provider `required-action` .
```
